Tolerate non-deterministic OpenJCEPlus signature

jasonkatonica opened the pull request:
The OpenJCEPlus signatures do not make use of random generators used in initialization routines. Instead they are ignored. This will cause test failures since the tests are expecting deterministic behavior instead of random behavior.

Signed-off-by: Jason Katonica <katonica@us.ibm.com>

jasonkatonica commented on test/jdk/sun/security/provider/DSA/SecureRandomReset.java:
```diff
@@ -20,7 +20,11 @@
+ * (c) Copyright IBM Corp. 2025, 2025 All Rights Reserved
+ * ===========================================================================
+ */
 /*
```
Added this blank line back after the first copyright statement.

keithc-ca replied:
> please add a blank line after this addition

This has not been addressed.

jasonkatonica replied:
Added a blank line after the IBM copyright statement.

jasonkatonica commented on test/jdk/ProblemList-FIPS140_3_OpenJcePlus.txt:
```diff
@@ -744,7 +744,6 @@ sun/security/pkcs12/StorePasswordTest.java https://github.com/eclipse-openj9/ope
 sun/security/pkcs12/StoreSecretKeyTest.java https://github.com/eclipse-openj9/openj9/issues/20978 generic-all
 sun/security/pkcs12/StoreTrustedCertTest.java https://github.com/eclipse-openj9/openj9/issues/20978 generic-all
 sun/security/pkcs12/WrongPBES2.java https://github.com/eclipse-openj9/openj9/issues/20978 generic-all
-sun/security/provider/DSA/SecureRandomReset.java https://github.com/eclipse-openj9/openj9/issues/20978 generic-all
```
I had at first tried to remove this test from the strict profile excludes list and realized that this would not work. The reason being that `DSA` key generators are not registered by the `OpenJCEPlusFIPS` provider. The test would fail with an `AlgorithmNotFound` exception on the first line of the test `KeyPairGenerator g = KeyPairGenerator.getInstance("DSA");`. When running in strict FIPS mode no other providers are available to find `DSA`.

keithc-ca replied:
Why was this reverted in f05e4057a829f69dff4d2da33924a7bce6e74be9?

keithc-ca replied:
Ah, I see https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/982#issuecomment-2779337297.

jasonkatonica commented on test/jdk/sun/security/provider/DSA/SecureRandomReset.java:
```diff
@@ -50,6 +54,16 @@ public static void main(String[] args) throws Exception {
+        // The OpenJCEPlus provider does not honor a secure random
+        // generator when used to initialize a signature.
+        if (s.getProvider().getName().equals("OpenJCEPlus")) {
+            if (Arrays.equals(sig1,sig2)) {
```
I added a space after the comma. Please see previous comment above regarding the FIPS provider to see if this comment now makes more sense.

keithc-ca replied:
If the exception does not apply to OpenJCEPlusFIPS, then the message on line 62 should not mention that provider.

jasonkatonica replied:
Yes agreed that message on line 62 was mistakenly left over from when i tried this with the fips provider. The mention of FIPS has been removed.

jasonkatonica commented:
Testing this with the weak FIPS profile recently resulted in the following exception. For this reason I also reverted the updates to the exclusion list that i made for weak profile such that this test is excluded in testing of both weak and strict profiles ( anything that uses OpenJCEPlusFIPS ) given this test is not possible to run in FIPS mode due to various DSA restrictions:
```
java.security.ProviderException: DSA signing not supported in FIPS
	at openjceplus/com.ibm.crypto.plus.provider.OpenJCEPlusFIPS.providerException(OpenJCEPlusFIPS.java:875)
	at openjceplus/com.ibm.crypto.plus.provider.DSASignature.engineInitSign(DSASignature.java:58)
	at java.base/java.security.SignatureSpi.engineInitSign(SignatureSpi.java:133)
	at java.base/java.security.Signature$Delegate.engineInitSign(Signature.java:1371)
	at java.base/java.security.Signature.initSign(Signature.java:659)
	at SecureRandomReset.main(SecureRandomReset.java:53)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:586)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:333)
	at java.base/java.lang.Thread.run(Thread.java:1595)
	```

keithc-ca commented:
What is the value in this change if the test remains excluded for all situations involving OpenJCEPlus providers?

jasonkatonica commented:
> What is the value in this change if the test remains excluded for all situations involving OpenJCEPlus providers?

The exclusion lists are associated with behavior for when the OpenJCEPlusFIPS provider is running first in the JCE provider list. While we are still excluding this test in both of our FIPS test configurations ( strict and weak profile ) the main value here is to tolerate the case where the non-FIPS OpenJCEPlus provider is running first in the JCE provider list. 

Our long term goal is to support users who want to choose to run OpenJCEPlus as their first provider in the list. 

In the even longer term plan we plan on shipping OpenJCEPlus as the default provider first in the Semeru provider list.

These kinds of test updates help us to be able to test these cases where non fips OpenJCEPlus is being used.

keithc-ca commented:
It remains excluded in the non-FIPS OpenJCEPlus case (by `ProblemList-FIPS140_3_OpenJcePlus.txt`), so I must not understand.

nit: Please hyphenate "non-determinist".

jasonkatonica commented:
> It remains excluded in the non-FIPS OpenJCEPlus case (by `ProblemList-FIPS140_3_OpenJcePlus.txt`), so I must not understand.

So the excludes file `ProblemList-FIPS140_3_OpenJcePlus.txt` represents all tests excluded when we run with weakly enforced FIPS mode. The weak FIPS profile places the `OpenJCEPlusFIPS` provider first in the JCE preference list however it does not remove various other providers from Sun from the list. 

The file `ProblemList-FIPS140_3_OpenJcePlus.FIPS140-3.txt` is also a FIPS exclusion list. This set of excludes is applied when we test with strict profile. In strict profile only true FIPS allowable by default algorithms are available. ( No cryptography from non fips sources like Sun )

> nit: Please hyphenate "non-determinist".

Updated PR title and also adjusted original underlying git commit text to use `non-deterministic`.

keithc-ca commented:
Ok, I was focusing on just the latter part of the names of those problem-list files; one ends with "OpenJcePlus.txt" and the other ends with "OpenJCEPlusFIPS.FIPS140-3.txt - I was reading into that that the first was for non-FIPS and the latter was for FIPS, when, as I now understand, they're both for FIPS-140-3 scenarios. Given the problem you described, we're not ready to unexclude the test in either of those files.

keithc-ca commented:
Please squash.

jasonkatonica commented:
> Please squash.

Squashed

keithc-ca commented:
> OpenJCEPlus and OpenJCEPlus signatures

Why is "OpenJCEPlus" repeated (in the description and the commit message)?

keithc-ca commented:
Jenkins test sanity plinux jdknext

keithc-ca reviewed (approved):
The test failures appear unrelated to this change.
